internal/session/sessionwantsender.go

```diff
@@ -247,6 +273,11 @@ func (sws *sessionWantSender) onChange(changes []change) {
 	// don't have the want
 	sws.checkForExhaustedWants(dontHaves, newlyUnavailable)
 
+	// If there are any cancels, send them
+	if cancels.Len() > 0 {
+		sws.canceller.CancelSessionWants(sws.sessionID, cancels.Keys())
```

Comment:
This can deadlock if the peermanager is blocked on sending changes to this service.

Reply:
Good catch - what do you think the best fix is?
We could either
- make a Canceller that has its own go-routine
- modify `SignalAvailability()` to be non-blocking (it would attempt to send on the changes channel, and if it's full it would launch a go-routine).

SignalAvailability() should be called infrequently, so I favour the approach of modifying SignalAvailability()

Reply:
Modifying SignalAvailability to not block sounds good.